Extract XMP data from embedded JPEG preview inside RAF files

cytrinox opened the pull request:
The Fujifilm X-T5 camera stores in-camera rating for RAF images by using XMP. But the XMP data is not directly encoded into the RAF structure - instead it is attached as a second APP1 segment to the embedded JPEG preview file.

This patch extracts the JPEG preview and parses it like a standalone JPEG file.

fixes #2610

cytrinox committed:
Extract XMP data from embedded JPEG preview inside RAF files

The Fujifilm X-T5 camera stores in-camera rating for RAF images
by using XMP. But the XMP data is not directly encoded into the
RAF structure - instead it is attached as a second APP1 segment
to the embedded JPEG preview file.

This patch extracts the JPEG preview and parses it like a standalone
JPEG file.

cytrinox commented:
To verify https://github.com/Exiv2/exiv2/issues/2610 one can use:

    exiv2 --Print IkytEX ../_DSF0052.RAF

The XMP rating is now printed:

    Xmp.xmp.Rating                               XmpText    4

cytrinox commented:
@kmilos Here we go. Maybe you can have a look at it - I've touched C++ code long time ago :smile: 

There is one thing that could be improved: The JpegImage metadata parser already parses exif and other information. There is no need to keep the old *direct* method by using TiffParser. But I don't know how the get the byte order for `setByteOrder(bo);`.

kmilos commented on src/rafimage.cpp:
```diff
@@ -248,17 +249,31 @@ void RafImage::readMetadata() {
+  // another APP1 container for XMP. Just use the JpegImage metadata parser for that.
+  if(xmpData_.empty()) {
+    BasicIo::UniquePtr jpg_io = std::make_unique<Exiv2::MemIo>(jpg_buf.data(), jpg_buf.size());
+    auto jpg_img = JpegImage(std::move(jpg_io), false);
```
Why not use this single decoder to extract both Exif and XMP in one shot?

I guess it might not be guaranteed that the above Exif APP1 segment assumed to be handled by TiffParser comes first...

cytrinox replied:
@kmilos Yes, thats my guess, too. And I don't know how I should get the byte order for `setByteOrder(bo);`.

kmilos replied:
You don't need it or care about it here. The JpegImage figures it out internally anyway.

kmilos replied:
I can maybe dig deeper into this, but not sure why `setByteOrder(bo)` is there to begin with... There is parsing of another Tiff structure in the RAF later, and that has it's own byte order (we don't know if it's guaranteed to be identical to the embedded JPEG one, but it most likely is given that it is serialized by the same firmware).

If you want to keep the same functionality, then you can just do something like `setByteOrder(jpg_img.byteOrder())` instead.

neheb commented on src/rafimage.cpp:
```diff
@@ -248,17 +249,31 @@ void RafImage::readMetadata() {
+    auto jpg_img = JpegImage(std::move(jpg_io), false);
+    jpg_img.readMetadata();
+    const auto jpg_xmp_data = jpg_img.xmpData();
+    for(auto it = jpg_xmp_data.begin(); it != jpg_xmp_data.end(); ++it) {
```
for range loop.

neheb commented on src/rafimage.cpp:
```diff
@@ -248,17 +249,31 @@ void RafImage::readMetadata() {
+  // If there is no XMP data in first APP1 EXIF structure, there is maybe
+  // another APP1 container for XMP. Just use the JpegImage metadata parser for that.
+  if(xmpData_.empty()) {
+    BasicIo::UniquePtr jpg_io = std::make_unique<Exiv2::MemIo>(jpg_buf.data(), jpg_buf.size());
```
auto

cytrinox committed:
RAF: Simplify metadata extraction

Remove manual TiffParser usage for reading APP1 segement,
instead use JpegImage for proper parsing.

kmilos commented:
Ah, probably want to wrap the whole JPEG parsing (incl. the two preview tags) in a try-catch now, or update the expected test output...

cytrinox committed:
RAF: Ignore JPEG decoder errors if the JPEG offsets are invalid

cytrinox commented:
> Ah, probably want to wrap the whole JPEG parsing (incl. the two preview tags) in a try-catch now, or update the expected test output...

fixed.

kmilos commented on src/rafimage.cpp:
```diff
@@ -248,23 +249,31 @@ void RafImage::readMetadata() {
+  } catch (const Exiv2::Error&) {
+  }
 
   exifData_["Exif.Image2.JPEGInterchangeFormat"] = getULong(jpg_img_offset, bigEndian);
   exifData_["Exif.Image2.JPEGInterchangeFormatLength"] = getULong(jpg_img_length, bigEndian);
```
These should be included in the `try {}` block above.

kmilos replied:
Hm, 50/50 on this one indeed...

Let's leave it as is then, and handle it separately if need be.

kmilos commented on src/rafimage.cpp:
```diff
@@ -248,23 +249,31 @@ void RafImage::readMetadata() {
+    auto jpg_img = JpegImage(std::move(jpg_io), false);
+    jpg_img.readMetadata();
+    setByteOrder(jpg_img.byteOrder());
+    xmpData_ = jpg_img.xmpData();
+    exifData_ = jpg_img.exifData();
+    iptcData_ = jpg_img.iptcData();
```
Copy over `comment_` as well.

kmilos replied:
That's right, an added benefit.

cytrinox replied:
Done.

kmilos commented:
Almost there! 👍

cytrinox committed:
RAF: Copy comment from embedded JPEG image

kmilos commented:
@Mergifyio backport 0.28.x
